packages/SwingSet/docs/vat-environment.md

```diff
@@ -0,0 +1,113 @@
+
+### Dynamic Import Expressions
+
+One active JS feature proposal would add a "dynamic import" expression: `await import('path')`. If the engine implements this, vat code might be able to bypass the Compartment's module map. For safety, the SES shim rejects code that looks like it uses this feature. The regular expression that looks for this pattern can be confused into falsely rejecting legitimate code, usually by having the word `import` in a comment at the end of a line.
+
+### Direct Eval Expressions
+
+A "direct eval", invoked like `eval(code)`, is defined to behave as if `code` were expanded in place. The evaluated code sees the same scope as the `eval` itself sees, so the `code` in:
+
+```js
+function foo(code) {
+  const x = 1;
+  eval(code);
+}
+```
+
+gets to reference `x`. If you perform a direct eval, you cannot hide your internal authorities from the code you're evaluating.
+
+In contrast, an "indirect eval" only gets the global scope, not the local scope. In a safe SES environment, indirect eval is a useful and common tool. The evaluated code can only access global objects, and those are all safe (and frozen). The only bad thing an indirect eval can do is consume unbounded CPU or memory. Once you've evaluated that code, you can give it as many or as few authories as you like, by invoking it with arguments.
+
+The most common way to invoke an indirect eval is like `(1,eval)(code)`.
+
+The SES shim cannot correctly emulate a "direct eval". If it tried, it would wind up performing an indirect eval. This could be pretty confusing, because the code might not actually use objects from the local scope, so you might not notice the problem until some future change altered the behavior.
+
+To avoid confusion, the shim use a regular expression to reject code that looks like it is performing a direct eval. This regexp is not complete (it is possible to trick it into performing a direct eval anyway), but that's ok, because our goal is merely to guide people away from confusing behaviors early in their development process.
+
+This regexp will falsely reject occurrences inside static strings and comments.
+
+### Frozen globalThis
+
+Vats run in a Compartment whose `globalThis` object is frozen. A mutable global object would provide an "ambient communication channel": one side sets `globalThis.heyBuddyAreYouOutThere = 'exfiltrated message'` and the other side periodically reads it. This would violate one of the rules of object-capability security: objects may only communicate through references.
+
+Vats can create new Compartments, and they get to decide whether the new `globalThis` is frozen or not. If the Compartment is meant to support ocap security, the creator should do `harden(c.globalThis)` before loading any code into it.
```

Comment:
`Compartment`s, and later `Compartment`

"whether" -> "if"

"is meant to support" -> "supports"

Reply:
The vat has the agency here, "if the `Compartment` supports ocap security" sounds (to me) like the vat creates something and then learns later whether it has a feature or not, outside of its control.

What about: "Vats can create new `Compartment`s, and they get to decide if the new compartment is meant to support ocap security or not. If so, they should run `harden(compartment.globalThis)` before loading any code into it." ?

Huh, you know, this section repeats an earlier paragraph.. I'll see if I can fix that.

Also, "Frozen globalThis" should not be part of the list of limitations/restrictions that result from using a shim rather than a native SES implementation. I think I'll change `## Other Changes` into `## Shim Limitations`, and promote the last three level-3 headings ("Frozen globalThis", "Frozen Primordials", and "Library Compatiblity" to level-2.

Reply:
Done, and I trimmed the earlier section with a pointer to this one. See what you think.